nuxeo-features/nuxeo-automation/nuxeo-automation-core/src/main/java/org/nuxeo/ecm/automation/core/util/JSONPropertyWriter.java

```diff
@@ -106,84 +242,108 @@ protected static void writeScalarPropertyValue(JsonGenerator jg, Property prop,
+            for (Property p : prop.getChildren()) {
+                // it's a list of complex object, don't write field names
+                writeProperty(jg, p, PropertyConsumer.nothing());
             }
         }
         jg.writeEndArray();
     }
 
-    protected static void writeMapPropertyValue(JsonGenerator jg, ComplexProperty prop, DateTimeFormat dateTimeFormat,
-            String filesBaseUrl) throws JsonGenerationException, IOException, PropertyException {
+    protected void writeMapProperty(JsonGenerator jg, ComplexProperty prop, PropertyConsumer fieldNameWriter)
+            throws PropertyException, JsonGenerationException, IOException {
+        if (!writeEmpty && (prop == null || prop.getChildren().isEmpty())) {
+            return;
+        }
+        fieldNameWriter.accept(jg, prop);
         jg.writeStartObject();
+        PropertyConsumer childFieldWriter = (j, p) -> j.writeFieldName(p.getName());
         for (Property p : prop.getChildren()) {
-            jg.writeFieldName(p.getName());
-            writePropertyValue(jg, p, dateTimeFormat, filesBaseUrl);
+            writeProperty(jg, p, childFieldWriter);
         }
         jg.writeEndObject();
     }
 
-    protected static void writeBlobPropertyValue(JsonGenerator jg, Property prop, String filesBaseUrl)
+    protected void writeBlobProperty(JsonGenerator jg, Property prop, PropertyConsumer fieldNameWriter)
             throws PropertyException, JsonGenerationException, IOException {
         Blob blob = (Blob) prop.getValue();
         if (blob == null) {
-            jg.writeNull();
-            return;
-        }
-        jg.writeStartObject();
-        String v = blob.getFilename();
-        if (v == null) {
-            jg.writeNullField("name");
-        } else {
-            jg.writeStringField("name", v);
-        }
-        v = blob.getMimeType();
-        if (v == null) {
-            jg.writeNullField("mime-type");
-        } else {
-            jg.writeStringField("mime-type", v);
-        }
-        v = blob.getEncoding();
-        if (v == null) {
-            jg.writeNullField("encoding");
-        } else {
-            jg.writeStringField("encoding", v);
-        }
-        v = blob.getDigest();
-        if (v == null) {
-            jg.writeNullField("digest");
+            if (writeNull) {
+                fieldNameWriter.accept(jg, prop);
+                jg.writeNull();
+            }
         } else {
-            jg.writeStringField("digest", v);
-        }
-        jg.writeNumberField("length", blob.getLength());
-        if (filesBaseUrl != null) {
-            jg.writeStringField("data", getBlobUrl(prop, filesBaseUrl));
+            fieldNameWriter.accept(jg, prop);
+            jg.writeStartObject();
+            String v = blob.getFilename();
+            if (v == null) {
+                if (writeNull) {
+                    jg.writeNullField("name");
+                }
+            } else {
+                if (writeNull) {
```

Comment:
Do you need this `if (writeNull)` here? It doesn't follow the pattern of the others below.